prompting/organic/organic_loop.py

```diff
@@ -0,0 +1,181 @@
+from starlette.types import Send
+from typing import Tuple, AsyncGenerator
+import time
+import json
+import asyncio
+from prompting.tasks.inference import InferenceTask
+import bittensor as bt
+from bittensor.dendrite import dendrite
+from prompting.rewards.scoring import task_scorer
+from prompting.datasets.base import ChatEntry
+from prompting.utils.uids import get_random_uids
+from prompting.base.forward import SynapseStreamResult
+from prompting.base.dendrite import DendriteResponseEvent
+from functools import partial
+
+
+class Completion:
+    uid: int
+    completed: bool = False
+    accumulated_chunks: list[str] = []
+    accumulated_chunks_timings: list[float] = []
+    accumulated_tokens_per_chunk: list[int] = []
+
+
+async def priority_fn(synapse: StreamPromptingSynapse) -> float:
+    """Priority function for the axon."""
+    return 10000000.0
+
+
+async def blacklist_fn(synapse: StreamPromptingSynapse) -> Tuple[bool, str]:
+    """Blacklist function for the axon."""
+    # ! DO NOT CHANGE `Tuple` return type to `tuple`, it will break the code (bittensor internal signature checks).
+    # We expect the API to be run with one specific hotkey (e.g. OTF).
+    return synapse.dendrite.hotkey != settings.ORGANIC_WHITELIST_HOTKEY, ""
+
+
+async def on_organic_entry(synapse: StreamPromptingSynapse) -> StreamPromptingSynapse:
+    """Organic query handle."""
+    logger.info(f"[Organic] Received from {synapse.dendrite.hotkey}, IP: {synapse.dendrite.ip}")
+    task = InferenceTask(query=synapse.messages[-1], messages=synapse.messages)
+    miner_synapse = StreamPromptingSynapse(
+        task_name=task.__class__.__name__,
+        seed=task.seed,
+        target_model=task.llm_model_id,
+        roles=["user"],
+        messages=task.messages,
+    )
+    # TODO: Query one of the top N incentive miners, keep the rest random.
+    uids = list(get_random_uids(k=settings.ORGANIC_SAMPLE_SIZE))
+    # completions: dict[int, dict] = {}
+    completions: list[Completion] = []
+    token_streamer = partial(
+        stream_miner_response,
+        miner_synapse,
+        uids,
+        completions,
+    )
+
+    streaming_response = miner_synapse.create_streaming_response(token_streamer)
+    logger.debug(f"Message: {miner_synapse.messages}; Completions: {completions}")
+    asyncio.create_task(wait_and_add(task, completions, synapse=synapse, uids=uids))
+    return streaming_response
+
+
+async def wait_and_add(
+    task: InferenceTask, completions: list[Completion], synapse: StreamPromptingSynapse, uids: list[int]
+):
+    logger.debug("[ORGANIC] Waiting for responses to be collected")
+    for _ in range(settings.ORGANIC_TIMEOUT):
+        logger.debug(f"[ORGANIC] {len(completions)} responses collected")
+
+        if len(completions) == len(uids):
+            break
+        await asyncio.sleep(1)
+    else:
+        logger.error("[ORGANIC] Some responses couldn't be collected in time...")
```

Comment:
asyncio.wait_for would be more readable
```suggestion
    async def wait_for_responses():
        while len(completions) < len(uids):
            await asyncio.sleep(0.1)

    try:
        await asyncio.wait_for(wait_for_responses(), timeout=settings.ORGANIC_TIMEOUT)
        logger.debug(f"[ORGANIC] All {len(uids)} responses collected successfully.")
    except asyncio.TimeoutError:
        logger.error("[ORGANIC] Some responses couldn't be collected in time...")
```